modules/styling/lib/cs.ts

```diff
@@ -0,0 +1,355 @@
+    ? `var(${input}, ${fallback.startsWith('--') ? `var(${fallback})` : fallback})`
+    : `var(${input})`;
+}
+
+/**
+ * Create temporary CSS variables to use in components. The CSS variable names will
+ * be unique at runtime to avoid collisions. The return value is a function and a
+ * Map. The function can be used to pass in values from JavaScript. The function will
+ * return a map of variable keys to CSS Variable names.
+ *
+ * ```ts
+ * // creates a `color` and `background` variable
+ * const myVars = createVars('color', 'background')
+ *
+ * // 'color' is a typed property. The type is `string`
+ * console.log(myValues.color) // `'var(--{hash}-color)'`
+ *
+ * // 'color' is a typed property. The type is `string?`
+ * // The returned object can be assigned to the `style` property of an element
+ * console.log(myVars({ color: 'red' })) //  `{'--{hash}-color': 'red'}`
+ *
+ * const div = document.createElement('div')
+ * div.style = myVars({ color: 'red' }) // <div style="--{hash}-color: red;" />
+ * ```
+ */
+export function createVars<T extends string, ID extends string>(input: {
+  id: ID;
+  args: T[];
+}): CsVars<T, ID>;
+export function createVars<T extends string>(...args: T[]): CsVars<T, never>;
+export function createVars<T extends string, ID extends string>(...args: T[]): CsVars<T, ID> {
+  const id = (args[0] as any).id || generateUniqueId(); //?
+  // eslint-disable-next-line no-param-reassign
+  args = (args[0] as any).args || args;
+  const result = (input: Record<T, string>) => {
+    const vars = {};
+    args.forEach(key => {
+      if (input[key]) {
+        // @ts-ignore TS complains about `key` not in object `{}`
+        vars[result[key]] = input[key];
+      }
+    });
+    return vars;
+  };
+
+  args.forEach(key => {
+    // @ts-ignore
+    result[key] = `--${id}-${slugify(key)}`;
+  }, {});
+
+  return result as CsVars<T, ID>;
+}
+
+type ModifierConfig = Record<string, Record<string, CS>>;
+
+type ModifierValues<T extends ModifierConfig> = {
+  [P in keyof T]: keyof T[P];
+};
+
+type ModifierReturn<T extends ModifierConfig> = T &
+  ((modifiers: Partial<ModifierValues<T>>) => string);
+
+/**
+ * Creates a modifier function that takes in a modifier config and will return a CSS class name that
+ * matches the result. Modifiers can be thought as `if` or `switch` statements. This function can be
+ * thought of as a helper function that makes it easier to work with modifiers. Without it, you
+ * would have to implement if/switch/ternary for each option.
+ *
+ * ```tsx
+ * const myModifiers = createModifiers({
+ *   // a modifier called 'size'
+ *   size: {
+ *     small: createStyles({ fontSize: 12 }),
+ *     medium: createStyles({ fontSize: 14 })
+ *   }
+ * })
+ *
+ * // with the modifier function
+ * myModifiers({ size: 'medium' }) // returns the medium class name
+ *
+ * // manually without the function
+ * size === 'small' ? myModifiers.size.small : size === 'medium' ? myModifiers.size.medium : ''
+ * ```
```

Comment:
i like the HTML reference that you have above, can we add something like:
```
// When passing the modifier of size to your component
<Component size="small" ...></Component>
// internal component code
<div className="default-classname modifier-small className"></div>
```
Very similar to how [BEM](https://getbem.com/introduction/) does it.

I'm ok if this doesn't live in the jsdoc but somewhere else

Reply:
Also probably worth noting that the order here is really important